### ISSUE TITLE
Ephemeral code evaluation

### ISSUE THREAD

**Issue by zmosheSC** [googlecolab/colabtools#2521]
**Describe the solution you'd like**

As a next-step to the "run-selected-text" feature, I think it could be very useful to have something like  "run-temporarily-modified-text". 

I'm using the run-selected-text feature quite a lot. Actually, can't imagine working without it. 
Many times though, I want to run a slightly modified version of the code in the cell, but don't want it to be persisted. For example - when I'm working with numpy/TF, it's very common to have a shape-mismatches in a long expression, and then I'd normally inspect at the variables one after another to spot the problem. In that case I don't really want to evaluate the variable itself, but only x.shape so I find myself editing the cell, adding ".shape" after different variables or methods invocations just to delete it right after.
Could be useful if in addition to Cmd-Shift-Enter for "run selected text", we'd have something similar that would open a textbox with the selected text, allows us to change it and run the modified version without saving it back to the cell itself.
(any other UX is of course possible, just wanted to get the idea).

**Describe alternatives you've considered**
Alternatives (which are what I do today) is to modify the cell itself, run the selected part, and then delete the changes. That works, but a but annoying.
I'm also aware of the "Inspector" tab, but this one only allows looking at variables and I can't easily chain methods invocation and inspect their results.

**Comment by blois** [comment]
If you go to View -> Executed Code History then it has a panel at the bottom which allows arbitrary expression evaluation which does not get persisted in the notebook.

The variable inspector and the hover tooltip should also show shape information- let us know if there's something missing from those.

**Comment by zmosheSC** [comment]
Yes! this is exactly what I meant functionally, just with a less intuitive UX.. 
(first, I'd never think of going to "code history" for something like this, but even when I know where it is, that's still a lot a copy-paste for the scenario where I'm re-evaluating parts of the code to figure out the problem).

So maybe I should rephrase the feature to a smaller one - can we define a command to be ("Show executed code history" + fill selected text in the inputbox)? Then we can assign a shortcut to it, making it useful.


As for the inspector and the tooltop - yes, they help in case it's a single variable but not when it's a series of functions invocations. See the following code for example (from the top of the head, doesn't do anything useful):
```
output = np.log(np.mean(input[None, ...], axis=-1) @ np.exp(weights))
```
when there's a mismatch somewhere, I'd probably want to look at `input`'s shape (easy, even with the tooltip), but then to check that the expansion worked (I currently copy to another row the `input[None,...]` part and evaluate only this), then the whole `np.mean()` expression, then the weights part and see if the shape matches for the matmul. 
Would be nice to select parts of the text, and get them in an evaluator where I can still modify before actually running (add '.shape' or count negative values before apply log, etc...).
